lib/sshkit.ex

```diff
@@ -262,17 +263,96 @@ defmodule SSHKit do
     Enum.map(context.hosts, run)
   end
 
-  # def upload(context, path, options \\ []) do
-  #   …
-  #   # resolve remote relative to context path
-  #   remote = Path.expand(Map.get(options, :as, Path.basename(path)), _)
-  #   SCP.upload(conn, path, remote, options)
-  # end
-
-  # def download(context, path, options \\ []) do
-  #   …
-  #   remote = _ # resolve remote relative to context path
-  #   local = Map.get(options, :as, Path.basename(path))
-  #   SCP.download(conn, remote, local, options)
-  # end
+  @doc ~S"""
+  Upload a file or files to the given context.
+  Returns a list of `:ok`.
+  There is one `:ok` per connected host where file upload was successful.
+
+  ## Examples
+
+  Upload all files and folders in current directory to "/workspace".
+
+  ```
+  [:ok] =
+    "my.remote-host.tld"
+    |> SSHKit.context
+    |> SSHKit.path("/workspace")
+    |> SSHKit.upload(".", recursive: true)
+
+  ```
+
+  Upload file to different name on host
+
+  ```
+  [:ok] =
+    "my.remote-host.tld"
+    |> SSHKit.context
+    |> SSHKit.upload("some_file.txt", as: "other_file.txt" )
+
+  ```
+
+  """
+  def upload(context, path, options \\ []) do
+    # resolve remote relative to context path
+    # override with :as option if present
+    destination_path = if (context.path), do: context.path, else: ""
+    destination = case Keyword.get(options, :as, false) do
+      false -> path
+      override -> override
+    end
+    remote = if (Path.type(destination) == :absolute), do: destination, else: Path.join(destination_path, destination)
+
+    run = fn host ->
+      {:ok, conn} = SSH.connect(host.name, host.options)
+      SCP.upload(conn, path, remote, options)
+    end
+
+    Enum.map(context.hosts, run)
+  end
```

Comment:
Hey @brienw, @tessi and me had a look at `upload` and came up with a bunch of ideas for refactoring it for better understanding and readability. We'd suggest to split off a `build_remote_path` function which can also be reused within `download`.

Here's what this could look like:

```elixir
def upload(context, path, options \\ []) do
  as_path = Keyword.get(options, :as, path)
  remote_path = build_remote_path(context, as_path)

  run = fn host ->
    {:ok, conn} = SSH.connect(host.name, host.options)
    res = SCP.upload(conn, path, remote_path, options)
    SSH.close(conn)
    res
  end

  Enum.map(context.hosts, run)
end

defp build_remote_path(context, path) do
 if Path.type(path) == :absolute
   path
 else
   Path.join(context.path || “”, path)
 end
end
```

`upload` now also closes the connection per host - something that for now may seem a bit tedious, but will probably be addressed when it comes to #41 ✌️ 

We'd also change the documentation to this:

    @doc ~S"""
    Upload one or many files to the hosts of the given context.

    Returns a list of upload-result-tuples -- one for each host.
    Each result tuple is either `{:ok, output, status}` or {:error, “reason”}.

    Possible options are `as: "new_name"` (specify the name of the uploaded file/directory)
    and all options specified in `SSHKit.SCP.Upload.transfer/4`.

    ## Examples

    Upload all files and folders in the current directory
    to the remote directory “/workspace”:
       
    ```
    [{:ok, _, _}] =
      "my.remote-host.tld"
      |> SSHKit.context
      |> SSHKit.path("/workspace")
      |> SSHKit.upload(".", recursive: true)
    ```

    Uploads a file with a different name to the target host:

    ```
    [{:ok, _, _}] =
      "my.remote-host.tld"
      |> SSHKit.context
      |> SSHKit.upload("some_file.txt", as: "other_file.txt" )
    ```

    """

It especially changes the part on what is returned by `upload`. We *think* 🙈  that this would be a list of either `{:ok, output, return_value}` or `{:error, "reason"}` tuples rather then only a list of `:ok` atoms.
Is this assumption correct?